### PR TITLE
Include tinyformat as a subtree

### PR DESCRIPTION
**Pull request by Empact**
This adds the existing c42f/tinyformat@689695c of tinyformat as a subtree, and extracts the modifications that had been made to a utilstrprintf.h wrapper file which is included instead of tinyformat.h directly.

The benefits being that this should be easier to verifiably update and clearer what changes are being introduced, while making it easier to identify tinyformat.h as a dependency rather than a project file, and apply special treatment relative to that fact.

This re-opens #13842

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK
I didn't do this initially because I didn't expect tinyformat to be an actually maintained library (as well as not really understanding submodules very well), but in retrospect this is better.

**Comment by scravy** [comment]
utACK – LGTM

**Comment by laanwj** [comment]
labeling this for after the 0.18 branch-off, I don't think this makes sense to do last-minute for 0.17

**Comment by Empact** [comment]
@laanwj Agreed

**Comment by Empact** [comment]
Repeating @MarcoFalke's concern here for visibility:
"Seems like a lot of files to pull in just to make it a subtree, no? Imo tinyformat is really meant to be a single file just to be copied into the project."
https://github.com/bitcoin/bitcoin/pull/13842#issuecomment-409732591

I have a mild preference for automatic checks of consistency with upstream, but #13846 exists as an alternative.

**Review by ajtowns — COMMENTED** [review]
Concept ACK

**Comment by laanwj** [comment]
> Seems like a lot of files to pull in just to make it a subtree, no? Imo tinyformat is really meant to be a single file just to be copied into the project."

After seeing the number of files added here I think I'm starting to agree with @MarcoFalke.

**Comment by Empact** [comment]
Here are the files included in the subtree - 10 in total: https://github.com/Empact/bitcoin/tree/tinyformat-subtree/src/tinyformat

The other changed files relate mostly to redirecting tinyformat.h includes to utilstrprintf.h.

**Comment by ajtowns** [comment]
The two PRs is a bit of a pain. I think I have a mild preference for the 10-files-but-automatic-checks over 2-files-but-manual-checks, but am not really bothered either way.

The 46-files-changed is mostly just churn to pull out the customisations into utilstrprintf.h. You could avoid that by renaming utilstrprintf.h back to tinyformat.h (which then includes tinyformat/tinyformat.h). If I'm counting right, doing that gets it down to 24 changes: 10 src/tinyformat/, src/tinyformat.h, dev-notes.md, copyright_header.py, Makefile.am, src/Makefile.am, 6 lint files, travis.yml, and src/primitives/block.{h,cpp}. The block.{h,cpp} changes are just to directly include some standard headers so could also be skipped.

File: src/Makefile.am [hunk 1]
```diff
@@ -182,6 +182,7 @@ BITCOIN_CORE_H = \
   util.h \
   utilmemory.h \
   utilmoneystr.h \
+  utilstrprintf.h \
```
Comment (ajtowns) [inline src/Makefile.am]: Is this change needed/useful?

**Comment by Empact** [comment]
One option is to split this into 2 prs: the extraction of utilstrprintf.h and the move to a subdir (which would be pretty trivial at that point).

**Comment by ken2812221** [comment]
Agree with @ajtowns, rename `utilstrprintf.h` to `tinyformat.h`

**Comment by laanwj** [comment]
Closing (see https://github.com/bitcoin/bitcoin/pull/13846#issuecomment-417666596)
